spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,149 @@
+
+A *user*, as referred to in this specification, is an entity capable of executing datagrams who is paying for computation / storage (via gas or a similar mechanism) but is otherwise untrusted. Possible users include:
+- End users signing with an account key
+- On-chain smart contracts acting autonomously or in response to another transaction
+- On-chain modules acting in response to another transaction or in a scheduled manner
+
+#### Requirements
+
+Connection handlers and subsequent protocols make use of a simple key-value store interface provided by the underlying state machine. This store must provide two functions, which behave in the way you would expect:
+- `Get(Key) -> Value | null`
+- `Set(Key, Value)`
+
+`Key` and `Value` are assumed to be byte slices; encoding details are left to a later ICS.
+
+Blockchains also need the ability to introspect their own root-of-trust (with `getRootOfTrust`) in order to confirm that the connecting chain has stored the correct one.
+
+#### Subprotocols
+
+Subprotocols are defined as a set of datagram types and a `handleDatagram` function which must be implemented by the state machine of the implementing blockchain. Datagrams must be relayed between chains by an external process. This process is assumed to behave in an arbitrary manner — no safety properties are dependent on its behavior, although progress is generally dependent on the existence of at least one correct relayer process. Further discussion is deferred to [ICS 18: Relayer Algorithms](../spec/ics-18-relayer-algorithms).
+
+IBC subprotocols are reasoned about as interactions between two chains `A` and `B` — there is no prior distinction between these two chains and they are assumed to be executing the same, correct IBC protocol. `A` is simply by convention the chain which goes first in the subprotocol and `B` the chain which goes second. Protocol definitions should generally avoid including `A` and `B` in variable names to avoid confusion (as the chains themselves do not know whether they are `A` or `B` in the protocol).
+
+This ICS defines four subprotocols: opening handshake, header tracking, closing handshake, and closing by equivocation.
+
+##### Opening Handshake
+
+The opening handshake subprotocol serves to initialize roots of trust for two chains on each other and negotiate an agreeable connection version.
+
+This subprotocol need not be permissioned, modulo anti-spam measures.
+
+![Opening Handshake](opening_handshake.png)
```

Comment:
Can we have a simple description of the fields of each datagram, either in or out of the latex block?

Reply:
Also might be better if we describe the direction
* `ConnOpenInit`: User -> chain A, 
* `ConnOpenTry`: chain A -> chain B
* `ConnOpenAck`: chain B -> chain A
* `ConnOpenConfirm` chain A -> chain B

Reply:
I've added a table describing the execution flow, great idea.

Also adding field descriptions.

Reply:
Added field descriptions in the LaTeX. Are they clear?